File: C++/soln-stack-problems/evaluate-reverse-polish-notation.cpp
/**
 @author Farheen Bano
  
 References-
 https://www.interviewbit.com/problems/evaluate-expression/
 https://leetcode.com/problems/evaluate-reverse-polish-notation/
*/

class Solution {
public:
    bool isOperator(string c){
        if(c=="+"||c=="-"||c=="/"||c=="*")
            return true;
        return false;
    }
    
    int evalRPN(vector<string>& tokens) {
        stack<int>st;
        int n = tokens.size();
        for(string token:tokens){
            if(!isOperator(token)){
                st.push(stoi(token));
            }
            else{
                int a = st.top();
                st.pop();
                int b = st.top();
                st.pop();
                if(token=="+"){
                    st.push(b+a);
                }
                else if(token=="-"){
                    st.push(b-a);
                }
                else if(token=="/"){
                    st.push(b/a);
                }
                else{
                    st.push(b*a);
                }
            }
        }
        return st.top();
    }
};

File: Python/soln-stack-problems/evaluate-reverse-polish-notation.py
'''
 @author Farheen Bano
  
 Reference-
 https://leetcode.com/problems/evaluate-reverse-polish-notation/
'''

class Solution:
    def evalRPN(self, tokens: List[str]) -> int:
        operandStack=[]
        operators={'+','-','*','/'}
        
        for token in tokens:
            if token in operators:
                if operandStack:
                    y=int(operandStack.pop())
                if operandStack:
                    x=int(operandStack.pop())
                    
                match token:
                    case '*':
                        operandStack.append(x*y)     
                    case '/':
                        operandStack.append(int(x/y)) 
                    case '+':
                        operandStack.append(x+y)  
                    case '-':
                        operandStack.append(x-y)                        
                    
            else:
                operandStack.append(token)
                
        return operandStack[-1]

# Stack Solution
# Time:     O(n),       Iterates through tokens at most one time.
# Space:    O(n),       list / stack could contain all tokens.
